### PR TITLE
Implement Release Parsing, Decision Engine, and Downloading

### PR DESCRIPTION
**Pull request by Qstick**
Implement Parsing, Decision Engine, and Downloading. This PR should get automatic/manual search working and send nzb/torrents to the download client.

#### Database Migration
YES

#### Description
This PR adds the following:
- basic release parsing for albums,
- decision engine for albums, 
- manual search and download in UI
- download client support for albums
- history view in UI
- Queue view in UI

#### Todos
- [x] [Fix Tests Build]
- [x] Decision Engine Tests
- [x] Parser Tests
- [x] Download Client Tests
- [x] Download Pending Tests
- [x] Download Approval Tests
- [x] Tracked Download Tests
- [x] Cleanup Code + Fix Codacy Issues

### CODE REVIEW

**Comment by galli-leo** [comment]
I skimmed it a bit and it seems good. I would add some Parser Tests though. From my experience with Radarr: you can never have enough tests for the parser :)

**Comment by Qstick** [comment]
Thanks @galli-leo, will do

File: src/NzbDrone.Core.Test/DecisionEngineTests/PrioritizeDownloadDecisionFixture.cs [hunk 1]
```diff
@@ -174,85 +145,52 @@ public void should_put_usenet_above_torrent_when_usenet_is_preferred()
...
+            decisions.Add(new DownloadDecision(remoteAlbum2));
 
             var qualifiedReports = Subject.PrioritizeDecisions(decisions);
-            qualifiedReports.First().RemoteEpisode.Episodes.Count.Should().Be(remoteEpisode1.Episodes.Count);
+            qualifiedReports.First().RemoteAlbum.Release.DownloadProtocol.Should().Be(DownloadProtocol.Torrent);
         }
 
         [Test]
-        public void should_prefer_single_episode_over_multi_episode_for_non_anime()
+        public void should_prefer_single_album_over_multi_album_for_non_anime()
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/PrioritizeDownloadDecisionFixture.cs]: But what about anime music?!

File: src/NzbDrone.Core.Test/DecisionEngineTests/ReleaseRestrictionsSpecificationFixture.cs [hunk 2]
```diff
@@ -106,13 +106,13 @@ public void should_ignore_case_when_matching_ignored(string ignored)
...
 
-            Subject.IsSatisfiedBy(_remoteEpisode, null).Accepted.Should().BeFalse();
+            Subject.IsSatisfiedBy(_remoteAlbum, null).Accepted.Should().BeFalse();
         }
 
         [Test]
         public void should_be_false_when_release_contains_one_restricted_word_and_one_required_word()
         {
-            _remoteEpisode.Release.Title = "[ www.Speed.cd ] -Whose.Line.is.it.Anyway.US.S10E24.720p.HDTV.x264-BAJSKORV";
+            _remoteAlbum.Release.Title = "[ www.Speed.cd ] -Whose.Line.is.it.Anyway.US.S10E24.720p.HDTV.x264-BAJSKORV";
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/ReleaseRestrictionsSpecificationFixture.cs]: Should we change release name?

File: src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/DelaySpecificationFixture.cs [hunk 3]
```diff
@@ -53,30 +53,32 @@ public void Setup()
...
-            _remoteEpisode.Release.DownloadProtocol = DownloadProtocol.Usenet;
+            _remoteAlbum.ParsedAlbumInfo = new ParsedAlbumInfo();
+            _remoteAlbum.Release = new ReleaseInfo();
+            _remoteAlbum.Release.DownloadProtocol = DownloadProtocol.Usenet;
 
-            _remoteEpisode.Episodes = Builder<Episode>.CreateListOfSize(1).Build().ToList();
-            _remoteEpisode.Episodes.First().EpisodeFileId = 0;
+            _remoteAlbum.Albums = Builder<Album>.CreateListOfSize(1).Build().ToList();
+
+            // TODO Rework album for track files
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/DelaySpecificationFixture.cs]: Does this needs to be handled?

File: src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/DelaySpecificationFixture.cs [hunk 4]
```diff
@@ -53,30 +53,32 @@ public void Setup()
...
-                  .Setup(s => s.GetPendingRemoteEpisodes(It.IsAny<int>()))
-                  .Returns(new List<RemoteEpisode>());
+                  .Setup(s => s.GetPendingRemoteAlbums(It.IsAny<int>()))
+                  .Returns(new List<RemoteAlbum>());
         }
 
         private void GivenExistingFile(QualityModel quality)
         {
-            _remoteEpisode.Episodes.First().EpisodeFileId = 1;
+            //_remoteAlbum.Albums.First().EpisodeFileId = 1;
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/DelaySpecificationFixture.cs]: Uncomment or remove?

File: src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/ProperSpecificationFixture.cs [hunk 5]
```diff
@@ -21,38 +21,43 @@ namespace NzbDrone.Core.Test.DecisionEngineTests.RssSync
...
             Mocker.Resolve<QualityUpgradableSpecification>();
 
-            _firstFile = new EpisodeFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 1)), DateAdded = DateTime.Now };
-            _secondFile = new EpisodeFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 1)), DateAdded = DateTime.Now };
+            _firstFile = new TrackFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 1)), DateAdded = DateTime.Now };
+            _secondFile = new TrackFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 1)), DateAdded = DateTime.Now };
 
-            var singleEpisodeList = new List<Episode> { new Episode { EpisodeFile = _firstFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = null } };
-            var doubleEpisodeList = new List<Episode> { new Episode { EpisodeFile = _firstFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = _secondFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = null } };
+            // TODO Rework for TrackFiles in Album
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/RssSync/ProperSpecificationFixture.cs]: Should this be done before merge?

File: src/NzbDrone.Core.Test/DecisionEngineTests/Search/ArtistSpecificationFixture.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        public void Setup()
+        {
+            _artist1 = Builder<Artist>.CreateNew().With(s => s.Id = 1).Build();
+            _artist2 = Builder<Artist>.CreateNew().With(s => s.Id = 2).Build();
+
+            _remoteAlbum.Artist = _artist1;
+        }
+
+        [Test]
+        public void should_return_false_if_series_doesnt_match()
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/Search/ArtistSpecificationFixture.cs]: Series -> album?

File: src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs [hunk 7]
```diff
@@ -21,39 +21,43 @@ public class UpgradeDiskSpecificationFixture : CoreTest<UpgradeDiskSpecification
...
             _upgradeDisk = Mocker.Resolve<UpgradeDiskSpecification>();
 
-            _firstFile = new EpisodeFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 2)), DateAdded = DateTime.Now };
-            _secondFile = new EpisodeFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 2)), DateAdded = DateTime.Now };
+            _firstFile = new TrackFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 2)), DateAdded = DateTime.Now };
+            _secondFile = new TrackFile { Quality = new QualityModel(Quality.MP3_512, new Revision(version: 2)), DateAdded = DateTime.Now };
 
-            var singleEpisodeList = new List<Episode> { new Episode { EpisodeFile = _firstFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = null } };
-            var doubleEpisodeList = new List<Episode> { new Episode { EpisodeFile = _firstFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = _secondFile, EpisodeFileId = 1 }, new Episode { EpisodeFile = null } };
+            //var singleEpisodeList = new List<Album> { new Album { EpisodeFile = _firstFile, EpisodeFileId = 1 }, new Album { EpisodeFile = null } };
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs]: Uncomment or remove?

File: src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs [hunk 8]
```diff
@@ -68,50 +72,51 @@ private void WithSecondFileUpgradable()
         }
 
         [Test]
-        public void should_return_true_if_episode_has_no_existing_file()
+        public void should_return_true_if_album_has_no_existing_file()
         {
-            _parseResultSingle.Episodes.ForEach(c => c.EpisodeFileId = 0);
+            //TODO Add for Albums
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs]: Does this need to be done before merging?

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/TorrentBlackholeFixture.cs [hunk 9]
```diff
@@ -185,7 +185,7 @@ public void Download_should_replace_illegal_characters_in_title()
             var illegalTitle = "Saturday Night Live - S38E08 - Jeremy Renner/Maroon 5 [SDTV]";
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/TorrentBlackholeFixture.cs]: Do these need to be changed? Magnet URI may need to be changed as well if it links to this torrent.

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/UsenetBlackholeFixture.cs [hunk 10]
```diff
@@ -119,7 +119,7 @@ public void Download_should_replace_illegal_characters_in_title()
             var illegalTitle = "Saturday Night Live - S38E08 - Jeremy Renner/Maroon 5 [SDTV]";
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/UsenetBlackholeFixture.cs]: Does this need to be changed to album titles?

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs [hunk 11]
```diff
@@ -41,15 +41,15 @@ public void Setup()
                     Id = RandomNumber,
                     DownloadedSize = 1000,
                     TotalDownloadSize = 10,
-                    GroupName = "tv",
+                    GroupName = "Music",
                     UiTitle = "Droned.S01E01.Pilot.1080p.WEB-DL-DRONE"
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs]: Does this need to be a track name?

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs [hunk 12]
```diff
@@ -41,15 +41,15 @@ public void Setup()
...
                 };
 
             _failed = new NzbVortexQueueItem
                 {
                     DownloadedSize = 1000,
                     TotalDownloadSize = 1000,
-                    GroupName = "tv",
+                    GroupName = "Music",
                     UiTitle = "Droned.S01E01.Pilot.1080p.WEB-DL-DRONE",
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs]: Track name?

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs [hunk 13]
```diff
@@ -59,9 +59,9 @@ public void Setup()
...
-                    DestinationPath = "/remote/mount/tv/Droned.S01E01.Pilot.1080p.WEB-DL-DRONE",
+                    DestinationPath = "/remote/mount/music/Droned.S01E01.Pilot.1080p.WEB-DL-DRONE",
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs]: Track name?

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbgetTests/NzbgetFixture.cs [hunk 14]
```diff
@@ -30,23 +30,23 @@ public void Setup()
...
                                               RecentTvPriority = (int)NzbgetPriority.High
                                           };
 
             _queued = new NzbgetQueueItem
                 {
                     FileSizeLo = 1000,
                     RemainingSizeLo = 10,
-                    Category = "tv",
+                    Category = "music",
                     NzbName = "Droned.S01E01.Pilot.1080p.WEB-DL-DRONE",
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbgetTests/NzbgetFixture.cs]: Track name? Also later in this file.

File: src/NzbDrone.Core.Test/Download/TrackedDownloads/TrackedDownloadServiceFixture.cs [hunk 15]
```diff
@@ -25,8 +26,8 @@ private void GivenDownloadHistory()
                  new History.History(){
                      DownloadId = "35238",
                      SourceTitle = "TV Series S01",
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/Download/TrackedDownloads/TrackedDownloadServiceFixture.cs]: Should this be an artist name?

File: src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs [hunk 16]
```diff
@@ -57,84 +54,62 @@ private int CompareAll(params int[] comparers)
...
-            }
-
-            return CompareByReverse(x.RemoteEpisode, y.RemoteEpisode, remoteEpisode => remoteEpisode.Episodes.Count);
-        }
-
-        private int CompareEpisodeNumber(DownloadDecision x, DownloadDecision y)
-        {
-            return CompareByReverse(x.RemoteEpisode, y.RemoteEpisode, remoteEpisode => remoteEpisode.Episodes.Select(e => e.EpisodeNumber).MinOrDefault());
-        }
+        // TODO Add compare track count if we can get the information from parsing release.
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs]: Should this be done before merge?

File: src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs [hunk 17]
```diff
@@ -159,7 +134,7 @@ private int CompareSize(DownloadDecision x, DownloadDecision y)
         {
             // TODO: Is smaller better? Smaller for usenet could mean no par2 files.
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs]: Done before merge?

File: src/NzbDrone.Core/DecisionEngine/DownloadDecisionMaker.cs [hunk 18]
```diff
@@ -120,30 +109,34 @@ private IEnumerable<DownloadDecision> GetDecisions(List<ReleaseInfo> reports, Se
...
+                var result = spec.IsSatisfiedBy(remoteAlbum, searchCriteriaBase);
 
                 if (!result.Accepted)
                 {
                     return new Rejection(result.Reason, spec.Type);
                 }
             }
+            catch (NotImplementedException e)
+            {
+                _logger.Trace("Spec " + spec.GetType().Name + " does not care about movies.");
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/DownloadDecisionMaker.cs]: movies -> music?

File: src/NzbDrone.Core/DecisionEngine/Specifications/AcceptableSizeSpecification.cs [hunk 19]
```diff
@@ -47,16 +42,17 @@ public Decision IsSatisfiedBy(RemoteEpisode subject, SearchCriteriaBase searchCr
             {
                 var minSize = qualityDefinition.MinSize.Value.Megabytes();
 
+                //TODO: implement if we can get album duractions from metadata
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/AcceptableSizeSpecification.cs]: Do before merge? Another todo later in file.

File: src/NzbDrone.Core/DecisionEngine/Specifications/CutoffSpecification.cs [hunk 20]
```diff
@@ -18,26 +19,30 @@ public CutoffSpecification(QualityUpgradableSpecification qualityUpgradableSpeci
...
-                
-                if (!_qualityUpgradableSpecification.CutoffNotMet(subject.Series.Profile, file.Quality, subject.ParsedEpisodeInfo.Quality))
-                {
-                    _logger.Debug("Cutoff already met, rejecting.");
-                    return Decision.Reject("Existing file meets cutoff: {0}", subject.Series.Profile.Value.Cutoff);
-                }
-            }
-
-            return Decision.Accept();
+            // TODO: Implement Cutoff checks for albums, require us to check tracks in album
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/CutoffSpecification.cs]: Do before merge?

File: src/NzbDrone.Core/DecisionEngine/Specifications/DiscographySpecification.cs [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+            _logger = logger;
+        }
+
+        public RejectionType Type => RejectionType.Permanent;
+
+        public virtual Decision IsSatisfiedBy(RemoteAlbum subject, SearchCriteriaBase searchCriteria)
+        {
+            throw new NotImplementedException();
+
+            //    if (subject.ParsedEpisodeInfo.FullSeason)
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/DiscographySpecification.cs]: Remove or uncomment?

File: src/NzbDrone.Core/DecisionEngine/Specifications/RawDiskSpecification.cs [hunk 22]
```diff
@@ -21,24 +21,26 @@ public RawDiskSpecification(Logger logger)
...
             {
                 return Decision.Accept();
             }
 
-            if (_dvdContainerTypes.Contains(subject.Release.Container.ToLower()))
-            {
-                _logger.Debug("Release contains raw DVD, rejecting.");
-                return Decision.Reject("Raw DVD release");
-            }
+            // TODO Implement for Music Releases if applicable.
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/RawDiskSpecification.cs]: Should this be done before merge?

File: src/NzbDrone.Core/DecisionEngine/Specifications/RssSync/DelaySpecification.cs [hunk 23]
```diff
@@ -49,38 +49,39 @@ public virtual Decision IsSatisfiedBy(RemoteEpisode subject, SearchCriteriaBase
...
-
-                        if (revisionUpgrade)
-                        {
-                            _logger.Debug("New quality is a better revision for existing quality, skipping delay");
-                            return Decision.Accept();
-                        }
-                    }
-                }
-            }
+            //TODO Rework for Albums
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/RssSync/DelaySpecification.cs]: Should this be done before merge?

File: src/NzbDrone.Core/DecisionEngine/Specifications/RssSync/ProperSpecification.cs [hunk 24]
```diff
@@ -22,32 +22,36 @@ public ProperSpecification(QualityUpgradableSpecification qualityUpgradableSpeci
...
-                    {
-                        _logger.Debug("Auto downloading of propers is disabled");
-                        return Decision.Reject("Proper downloading is disabled");
-                    }
-                }
-            }
-
-            return Decision.Accept();
+            throw new NotImplementedException();
+
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/RssSync/ProperSpecification.cs]: Should this be done before merge?

File: src/NzbDrone.Core/DecisionEngine/Specifications/SameEpisodesGrabSpecification.cs [hunk 25]
```diff
@@ -17,15 +18,17 @@ public SameEpisodesGrabSpecification(SameEpisodesSpecification sameEpisodesSpeci
...
         {
-            if (_sameEpisodesSpecification.IsSatisfiedBy(subject.Episodes))
-            {
-                return Decision.Accept();
-            }
+            throw new NotImplementedException();
 
-            _logger.Debug("Episode file on disk contains more episodes than this release contains");
-            return Decision.Reject("Episode file on disk contains more episodes than this release contains");
+            //    if (_sameEpisodesSpecification.IsSatisfiedBy(subject.Episodes))
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/SameEpisodesGrabSpecification.cs]: Uncomment or remove?

File: src/NzbDrone.Core/DecisionEngine/Specifications/Search/DailyEpisodeMatchSpecification.cs [hunk 26]
```diff
@@ -18,26 +19,28 @@ public DailyEpisodeMatchSpecification(Logger logger, IEpisodeService episodeServ
...
+        public Decision IsSatisfiedBy(RemoteAlbum subject, SearchCriteriaBase searchCriteria)
         {
-            if (searchCriteria == null)
-            {
-                return Decision.Accept();
-            }
+            throw new NotImplementedException();
 
-            var dailySearchSpec = searchCriteria as DailyEpisodeSearchCriteria;
+            //    if (searchCriteria == null)
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/Search/DailyEpisodeMatchSpecification.cs]: Uncomment or remove? Change filename?

File: src/NzbDrone.Core/DecisionEngine/Specifications/Search/SeasonMatchSpecification.cs [hunk 27]
```diff
@@ -15,23 +16,25 @@ public SeasonMatchSpecification(Logger logger)
...
         {
-            if (searchCriteria == null)
-            {
-                return Decision.Accept();
-            }
+            throw new NotImplementedException();
 
-            var singleEpisodeSpec = searchCriteria as SeasonSearchCriteria;
-            if (singleEpisodeSpec == null) return Decision.Accept();
+            //    if (searchCriteria == null)
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/Search/SeasonMatchSpecification.cs]: Uncomment or remove? Change file name?

File: src/NzbDrone.Core/DecisionEngine/Specifications/Search/SingleAlbumSearchMatchSpecification.cs [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+                return Decision.Accept();
+            }
+                
+            if (Parser.Parser.CleanArtistTitle(singleEpisodeSpec.AlbumTitle) != Parser.Parser.CleanArtistTitle(remoteAlbum.ParsedAlbumInfo.AlbumTitle))
+            {
+                _logger.Debug("Album does not match searched album title, skipping.");
+                return Decision.Reject("Wrong album");
+            }
+
+            //if (!remoteAlbum.ParsedAlbumInfo.AlbumTitles.Any())
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/Search/SingleAlbumSearchMatchSpecification.cs]: Uncomment or remove?

File: src/NzbDrone.Core/DecisionEngine/Specifications/UpgradeDiskSpecification.cs [hunk 29]
```diff
@@ -18,25 +19,28 @@ public UpgradeDiskSpecification(QualityUpgradableSpecification qualityUpgradable
...
-                _logger.Debug("Comparing file quality with report. Existing file is {0}", file.Quality);
-
-                if (!_qualityUpgradableSpecification.IsUpgradable(subject.Series.Profile, file.Quality, subject.ParsedEpisodeInfo.Quality))
-                {
-                    return Decision.Reject("Quality for existing file on disk is of equal or higher preference: {0}", file.Quality);
-                }
-            }
-
-            return Decision.Accept();
+            // TODO Refactor for Albums, once we have a way to check track quality by album.
```
Comment (danielunderwood) [inline src/NzbDrone.Core/DecisionEngine/Specifications/UpgradeDiskSpecification.cs]: Should this be done before merge?

File: src/NzbDrone.Core/Download/Clients/Pneumatic/Pneumatic.cs [hunk 30]
```diff
@@ -32,15 +32,16 @@ public class Pneumatic : DownloadClientBase<PneumaticSettings>
...
-            var title = remoteEpisode.Release.Title;
-
-            if (remoteEpisode.ParsedEpisodeInfo.FullSeason)
-            {
-                throw new NotSupportedException("Full season releases are not supported with Pneumatic.");
-            }
+            var url = remoteAlbum.Release.DownloadUrl;
+            var title = remoteAlbum.Release.Title;
+
+            // TODO Implement discographies
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Download/Clients/Pneumatic/Pneumatic.cs]: Should this be done before merge?

File: src/NzbDrone.Core/Download/DownloadFailedEvent.cs [hunk 31]
```diff
@@ -13,7 +13,9 @@ public DownloadFailedEvent()
         }
 
         public int SeriesId { get; set; }
+        public int ArtistId { get; set; }
         public List<int> EpisodeIds { get; set; }
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Download/DownloadFailedEvent.cs]: Remove or `System.Obsolete` episode stuff here?

File: src/NzbDrone.Core/Parser/Model/ParsedAlbumInfo.cs [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+namespace NzbDrone.Core.Parser.Model
+{
+    public class ParsedAlbumInfo
+    {
+        public string AlbumTitle { get; set; }
+        public string ArtistName { get; set; }
+        public ArtistTitleInfo ArtistTitleInfo { get; set; }
+        public QualityModel Quality { get; set; }
+        public string ReleaseDate { get; set; }
+        //public int[] TrackNumbers { get; set; }
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Parser/Model/ParsedAlbumInfo.cs]: Uncomment or remove?

File: src/NzbDrone.Core/Parser/Parser.cs [hunk 33]
```diff
@@ -39,6 +39,18 @@ public static class Parser
                     RegexOptions.IgnoreCase | RegexOptions.Compiled),
         };
 
+        private static readonly Regex[] ReportAlbumTitleRegex = new[]
+        {
+                //TODO Make more robust
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Parser/Parser.cs]: Does this need to be done before merging?

File: src/NzbDrone.Core/Parser/Parser.cs [hunk 34]
```diff
@@ -499,6 +511,110 @@ public static ParsedTrackInfo ParseMusicTitle(string title)
...
+                    title = new string(titleWithoutExtension) + title.Substring(titleWithoutExtension.Length);
+
+                    Logger.Debug("Reversed name detected. Converted to '{0}'", title);
+                }
+
+                var simpleTitle = SimpleTitleRegex.Replace(title, string.Empty);
+
+                simpleTitle = RemoveFileExtension(simpleTitle);
+
+                // TODO: Quick fix stripping [url] - prefixes.
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Parser/Parser.cs]: Does thie need to be done before merging?

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 35]
```diff
@@ -20,7 +20,10 @@ public interface IParsingService
         Series GetSeries(string title);
         RemoteEpisode Map(ParsedEpisodeInfo parsedEpisodeInfo, int tvdbId, int tvRageId, SearchCriteriaBase searchCriteria = null);
```
Comment (danielunderwood) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Can we remove these or `System.Obsolete` them?

File: src/NzbDrone.Core.Test/DecisionEngineTests/AcceptableSizeSpecificationFixture.cs [hunk 36]
```diff
@@ -171,50 +167,25 @@ public void should_return_true_if_unlimited_30_minute()
...
-            parseResultSingle.Series.SeriesType = SeriesTypes.Daily;
-            parseResultSingle.Release.Size = 300.Megabytes();
-
-            qualityType.MaxSize = 10;
-
-            Subject.IsSatisfiedBy(parseResultSingle, null).Accepted.Should().BeTrue();
-        }
-
-        [Test]
         public void should_return_true_if_RAWHD()
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/DecisionEngineTests/AcceptableSizeSpecificationFixture.cs]: I think this is checking for an actual full disc rip image (ISO, etc). I wasn't sure if we would run across that in music scene

File: src/NzbDrone.Core.Test/DecisionEngineTests/DownloadDecisionMakerFixture.cs [hunk 37]
```diff
@@ -118,59 +118,59 @@ public void should_have_same_number_of_rejections_as_specs_that_failed()
...
 
             results.Should().BeEmpty();
         }
 
         [Test]
-        public void should_not_attempt_to_map_episode_series_title_is_blank()
+        public void should_not_attempt_to_map_album_artist_title_is_blank()
         {
             GivenSpecifications(_pass1, _pass2, _pass3);
             _reports[0].Title = "1937 - Snow White and the Seven Dwarves";
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/DecisionEngineTests/DownloadDecisionMakerFixture.cs]: Yes

File: src/NzbDrone.Core.Test/DecisionEngineTests/ReleaseRestrictionsSpecificationFixture.cs (hunk 2 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/DecisionEngineTests/ReleaseRestrictionsSpecificationFixture.cs]: Yes

File: src/NzbDrone.Core.Test/DecisionEngineTests/Search/ArtistSpecificationFixture.cs (hunk 6 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/DecisionEngineTests/Search/ArtistSpecificationFixture.cs]: Will change

File: src/NzbDrone.Api/Indexers/ReleaseModule.cs [hunk 38]
```diff
@@ -26,6 +26,7 @@ public class ReleaseModule : ReleaseModuleBase
         private readonly Logger _logger;
 
         private readonly ICached<RemoteEpisode> _remoteEpisodeCache;
```
Comment (Qstick) [inline src/NzbDrone.Api/Indexers/ReleaseModule.cs]: Removed

File: src/NzbDrone.Api/Indexers/ReleaseResource.cs [hunk 39]
```diff
@@ -25,13 +25,10 @@ public class ReleaseResource : RestResource
         public string ReleaseGroup { get; set; }
         public string ReleaseHash { get; set; }
         public string Title { get; set; }
-        public bool FullSeason { get; set; }
-        public int SeasonNumber { get; set; }
         public Language Language { get; set; }
         public string AirDate { get; set; }
```
Comment (Qstick) [inline src/NzbDrone.Api/Indexers/ReleaseResource.cs]: Changed

File: src/NzbDrone.Core.Test/DecisionEngineTests/HistorySpecificationFixture.cs [hunk 40]
```diff
@@ -113,106 +112,106 @@ public void should_return_true_if_latest_history_item_is_not_grabbed()
         [Test]
         public void should_return_true_if_latest_history_item_is_older_than_twelve_hours()
         {
-            GivenMostRecentForEpisode(FIRST_EPISODE_ID, string.Empty, _notupgradableQuality, DateTime.UtcNow.AddHours(-13), HistoryEventType.Grabbed);
+            GivenMostRecentForAlbum(FIRST_ALBUM_ID, string.Empty, _notupgradableQuality, DateTime.UtcNow.AddHours(-13), HistoryEventType.Grabbed);
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/DecisionEngineTests/HistorySpecificationFixture.cs]: Changed to 12 hours 1 millisecond. Not a huge deal, but you are right this is more accurate test.

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/TorrentBlackholeFixture.cs (hunk 9 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/TorrentBlackholeFixture.cs]: Changed

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/UsenetBlackholeFixture.cs (hunk 10 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/Blackhole/UsenetBlackholeFixture.cs]: Changed

File: src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs (hunk 11 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/Download/DownloadClientTests/NzbVortexTests/NzbVortexFixture.cs]: Should be an album name as that's what we are grabbing. Ill fix

File: src/NzbDrone.Core.Test/Download/TrackedDownloads/TrackedDownloadServiceFixture.cs (hunk 15 above)
Comment (Qstick) [inline src/NzbDrone.Core.Test/Download/TrackedDownloads/TrackedDownloadServiceFixture.cs]: Fixed

File: src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs (hunk 17 above)
Comment (Qstick) [inline src/NzbDrone.Core/DecisionEngine/DownloadDecisionComparer.cs]: This is remnant of Sonarr TODOs

File: src/NzbDrone.Core/DecisionEngine/Specifications/RawDiskSpecification.cs (hunk 22 above)
Comment (Qstick) [inline src/NzbDrone.Core/DecisionEngine/Specifications/RawDiskSpecification.cs]: Not sure that there is such a thing as a raw rip for music releases, but I left it in encase it pops up. Removing blu-ray types and changing DVD to CD. Modified test fixture accordingly.

File: src/NzbDrone.Core/Parser/ParsingService.cs (hunk 35 above)
Comment (Qstick) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Will Obsolete them. Removing them kills many parser tests. Save that for another "Parser Refactor" PR.

File: src/NzbDrone.Core/Parser/Parser.cs (hunk 34 above)
Comment (Qstick) [inline src/NzbDrone.Core/Parser/Parser.cs]: Sonarr todo, will look at later

File: src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs [hunk 41]
```diff
@@ -30,35 +28,35 @@ public class UpgradeDiskSpecificationFixture : CoreTest<UpgradeDiskSpecification
...
+                ParsedAlbumInfo = new ParsedAlbumInfo { Quality = new QualityModel(Quality.MP3_256, new Revision(version: 2)) },
                 Albums = doubleEpisodeList
             };
 
             _parseResultSingle = new RemoteAlbum
             {
-                Artist = fakeSeries,
-                ParsedAlbumInfo = new ParsedAlbumInfo { Quality = new QualityModel(Quality.MP3_192, new Revision(version: 2)) },
+                Artist = fakeArtist,
+                ParsedAlbumInfo = new ParsedAlbumInfo { Quality = new QualityModel(Quality.MP3_256, new Revision(version: 2)) },
```
Comment (danielunderwood) [inline src/NzbDrone.Core.Test/DecisionEngineTests/UpgradeDiskSpecificationFixture.cs]: I noticed several of these changes between music qualities. They don't affect testing results or anything, right?

**Pull request merged**